### ISSUE TITLE
Add option to not include undefined values in resulting object literal

### ISSUE THREAD

**Issue by jchavarri** [ocsigen/js_of_ocaml#801]
Using

```ocaml
let obj = object%js (self)
  val z = Js.undefined [@@jsoo.optdef]
end
```

It still includes `{z: undefined}` in the resulting object literal. It'd be helpful to have an option (or a new attribute) to just not include the property altogether if it's undefined.

BuckleScript has something similar with `bs.obj` ([example](https://reasonml.github.io/en/try?rrjsx=true&reason=NoAQRgzgdA9mBWBdABAUwB4BdUCcB2AhgDbIDGOqB2AXMgBQBQyyAfgF7UCWemAvAPwAaJsgCueTpgYBKZLwB8yAPpzkAIjUBuBgyKpMyOPACMq8pWx0606dr0GjAJjMUqqOu16PB9G9qA)).

cc @Drup

**Comment by hhugo** [comment]
What would you expect the following to do ?
```ocaml
let f x =
  object%js (self)
    val z = x [@@jsoo.optdef]
  end
```

**Comment by jchavarri** [comment]
@hhugo I don't think the current behavior of `[@@jsoo.optdef]` is totally unexpected, but I was hoping to have a way for the property to not be included in the resulting literal if the value is undefined.

So in the snippet you shared, calling `f Js.undefined` would return an empty object `{}`.

**Comment by jchavarri** [comment]
In case it helps, [this](https://github.com/BuckleScript/bucklescript/blob/b55e6f24a0f18ce7844d5d4c05cdcf2ca69fccae/jscomp/syntax/ast_external_process.ml#L415-L517) is the part of the BuckleScript codebase where the `external ... = "" [@@bs.obj]` case is handled.

[Here's](https://bucklescript.github.io/bucklescript/Manual.html#_create_js_objects_using_external) the documentation for it.

As you can see, it walks through the args of the external declaration, and then calls `Ffi_obj_create`. The functionality provided by `external` + `[@@bs.obj]` has been extremely helpful in my experience, as it allows to create these literals with optional properties in a very ergonomic way using function labels.

I know this kind of design with annotations for externals is not common in jsoo... but I thought it'd be worth mentioning for more context.

**Comment by hhugo** [comment]
I'm not against the idea. What do you expect the generated javascript to be. Creating an empty object and and have a bunch of if-then to set fields if not undefined?

**Comment by jchavarri** [comment]
This is a sample implementation I did using the existing APIs:

```ocaml
let optInj prop opt =
  match opt with
  | Some (s) -> [|(prop, Js.Unsafe.inject s)|]
  | None  -> [||]

let create ?x:(x : string option)  ?y:(y : int option)  ?z:(z : int option) () =
  optInj "x" x
  |> Array.append (optInj "y" y)
  |> Array.append (optInj "z" z)

let obj = Js.Unsafe.obj (create ~x:"2" ~y:2 ()) (* {x: "2", y: 2} *)
let obj2 = Js.Unsafe.obj (create ()) (* {} *)
```

**Comment by jchavarri** [comment]
> What do you expect the generated javascript to be. Creating an empty object and and have a bunch of if-then to set fields if not undefined?

@hhugo Yes, in the same vein as [this](https://reasonml.github.io/en/try?rrjsx=true&reason=NoAQRgzgdA9mBWBdABAUwB4BdUCcB2AhgDbIDGOqB2AXMgBQBQyyAfgF7UCWemAvAPwAaJsgCueTpgYBKZLwB8yAPpzkAIjUBuBgyKpMyOPFXlK2bUA).

As a side question, I'm not sure how BuckleScript does it but it inlines most of the usages. I guess this is an optimization that already exists in the OCaml compiler, right?

**Comment by xguerin** [comment]
Was there any action taken as a result of this question? I would find that capability quite useful.
